add patch endpoint for editing workspace attributes

MatthewBemis opened the pull request:
This is part of the UI work that Blue Team has taken on for [DSDEEPB-1137](https://broadinstitute.atlassian.net/browse/DSDEEPB-1137).

eddiebroad commented:
it is my understanding that all orchestration PRs that include the implementation of a new service or pass-through require (by mutual agreement of group members) the addition of tests under /src/test/scala/org/broadinstitute/dsde/firecloud/.  I don't see such commits here

MatthewBemis commented:
I'll add those!

eddiebroad commented:
@MatthewBemis If you ask me, I would suggest adding a minimum of two tests.

eddiebroad commented:
@MatthewBemis Also, tests are required to pass before any merge is allowed

MatthewBemis commented:
@eddiebroad That was a mistake, but I've corrected that now. I replaced the two tests with just one because it's just a pass-through to rawls and (I think) 1 test should be sufficient to test that it works. I'm a bit unfamiliar with how your mocks work, but it looks like I can't test that an invalid attribute update operation returns Bad Request, because that's a response that rawls would provide, and rawls has it's own unit tests to check for this type of thing. Am I right in assuming this?

Also, I'll rebase again since there are new conflicts.

rushtong commented:
I'm looking at this now Matt, thanks!

rushtong commented on src/main/resources/configurations.conf:
```diff
@@ -25,4 +25,5 @@ workspace {
   submissionsPath="/workspaces/%s/%s/submissions"
   submissionByIdPath="/workspaces/%s/%s/submissions/%s"
   workflowOutputsByIdPath="/workspaces/%s/%s/submissions/%s/workflows/%s/outputs"
+  updateAttributesOnWorkspacePath="/workspaces/%s/%s/updateAttributes"
```
Which endpoint is using this path? I don't see any usage of it in the rest of the code.

rushtong commented on src/main/resources/swagger/api-docs.yaml:
```diff
@@ -599,6 +599,41 @@ paths:
+        - name: workspaceUpdateJson
+          description: Attribute operations
+          required: true
+          type: Array[AttributeUpdateOperation]
```
I'm pretty sure this isn't the correct syntax for the type. I think it is something like this, where `AttributeUpdateOperation` is defined in the definitions:
```
  type: array
    items:
      $ref: "#/definitions/AttributeUpdateOperation"
```

rushtong replied:
See spec here: http://swagger.io/specification/
Lots of examples of `type: array`

rushtong commented on src/main/resources/swagger/api-docs.yaml:
```diff
@@ -599,6 +599,41 @@ paths:
+  /workspaces/{workspaceNamespace}/{workspaceName}/updateAttributes:
+    get:
+      tags:
+      - Workspaces
```
This introduces a new header section in the swagger UI. Could you also update the other "workspaces" tags (lowercased) with the upper case version as part of this? It would be purely for the sake of consistency.

rushtong commented on src/test/scala/org/broadinstitute/dsde/firecloud/mock/MockWorkspaceServer.scala:
```diff
@@ -364,6 +368,20 @@ object MockWorkspaceServer {
+          .withHeaders(header)
+          .withStatusCode(OK.intValue)
+      )
+
```
It isn't too bad creating a negative test case. If you define a when case that listens for the values `mockInvalidWorkspace`, and then responds with what Rawls would respond with, you could then call that invalid case in your test class. 

If you did do that, you'd need to modify this when handler to only respond to a valid case, such as the values in `mockValidWorkspace` and call it appropriately in the test class.

rushtong commented on src/test/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceServiceSpec.scala:
```diff
@@ -41,6 +41,10 @@ class WorkspaceServiceSpec extends FreeSpec with ScalaFutures with ScalatestRout
 
+    val attributeUpdatePath = ApiPrefix + "/%s/%s/updateAttributes".format(
+      MockWorkspaceServer.mockValidWorkspace.namespace.get,
+      MockWorkspaceServer.mockValidWorkspace.name.get)
```
Here's where you'd add an invalid attributeUpdatePath using mockInvalidWorkspace.

rushtong commented on src/test/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceServiceSpec.scala:
```diff
@@ -185,6 +189,16 @@ class WorkspaceServiceSpec extends FreeSpec with ScalaFutures with ScalatestRout
+          status should equal(OK)
+        }
+      }
+    }
```
Here you would add the negative case that would return what Rawls would return as an error case.

rushtong commented:
@MatthewBemis - I'm done reviewing. I think there's a small bit of work left to do, specifically around the yaml and the test. Let me know when you'd like me to re-review.

rushtong commented on src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala:
```diff
@@ -86,6 +90,14 @@ trait WorkspaceService extends HttpService with PerRequestCreator with FireCloud
         } ~
+        path("updateAttributes") {
+          patch {
+            entity(as[Seq[AttributeUpdateOperation]]) { attributeOp =>
```
I am reconsidering my initial approval of this. I think we should just pass whatever entity is sent through as the entity without coercing it to `[Seq[AttributeUpdateOperation]]`. This will save you a bunch of code and simplify the passthrough. You can be more clear about what `AttributeUpdateOperation` is in the yaml, but I don't think we need this in the code.

MatthewBemis replied:
I'm almost done addressing the comments, but just a tiny bit confused on this part since I couldn't find any examples of it in the code. If we're just going to pass it through without caring about what it is so rawls can deal with it, what would be the entity type here? Would it be String? i.e.:

```
        path("updateAttributes") {
          patch {
            entity(as[String]) { attributeOp =>
              requestContext =>
                externalHttpPerRequest(requestContext, Patch(workspacePath, attributeOp))
            }
          }
        }
```
Or would it be some sort of wildcard type?

rushtong replied:
How about spray.json.JsObject ?

rushtong commented:
@MatthewBemis - Ignore my comments about adding a negative test case. There's no real value in the effort.

MatthewBemis commented:
@rushtong Ready for re-review!

rushtong commented:
:+1: Code looks good. I can't actually run the code, but I'm pretty sure that's a dev environment thing.

rushtong commented:
@MatthewBemis - Please squash and merge when complete.
